Frameworks/CoreText/DWriteWrapper_CoreText.mm

```diff
@@ -802,4 +803,135 @@ CFStringRef _DWriteFontCopyName(const ComPtr<IDWriteFontFace>& fontFace, CFStrin
+ * As such, this class is deliberately left thread-unsafe.
+ */
+class CGPathGeometrySink : public RuntimeClass<RuntimeClassFlags<WinRtClassicComMix>, IDWriteGeometrySink> {
+protected:
+    InspectableClass(L"Windows.Bridge.DirectWrite", TrustLevel::BaseTrust);
+
+public:
+    CGPathGeometrySink(const CGAffineTransform* transform) {
+        _cgPath.reset(CGPathCreateMutable());
+        _transform = transform;
+    }
+
+    HRESULT RuntimeClassInitialize() {
+        return S_OK;
+    }
+
+    void STDMETHODCALLTYPE AddBeziers(_In_ const D2D1_BEZIER_SEGMENT* beziers, unsigned int beziersCount) {
+        // Note that both AddBeziers and CGPathAddCurveToPoint use cubic Beziers (rather than quadratic)
+        for (unsigned int i = 0; i < beziersCount; ++i) {
+            CGPathAddCurveToPoint(_cgPath.get(),
+                                  _transform,
+                                  beziers[i].point1.x,
+                                  -beziers[i].point1.y,
+                                  beziers[i].point2.x,
+                                  -beziers[i].point2.y,
+                                  beziers[i].point3.x,
+                                  -beziers[i].point3.y);
+        }
+    }
+
+    void STDMETHODCALLTYPE AddLines(_In_ const D2D1_POINT_2F* points, unsigned int pointsCount) {
+        // Use individual CGPathAddLineToPoint() calls here, rather than CGPathAddLines
+        // CGPathAddLines does a CGPathMoveToPoint to the first point,
+        // which doesn't match what DWrite expects (draw a line from the previous point)
+        for (unsigned int i = 0; i < pointsCount; ++i) {
+            CGPathAddLineToPoint(_cgPath.get(), _transform, points[i].x, -points[i].y);
+        }
+    }
+
+    void STDMETHODCALLTYPE BeginFigure(D2D1_POINT_2F startPoint, D2D1_FIGURE_BEGIN figureBegin) {
+        if (_figureInProgress) {
+            TraceError(TAG,
+                       L"IDWriteGeometrySink::BeginFigure called while a figure was currently in progress. Placing object in error state - "
+                       L"future function calls will fail.");
+            _invalidState = true;
+        }
+
+        if (!_invalidState) {
+            // figureBegin is ignored for CGPath purposes -
+            // filled vs hollow maps to CGPathDrawingMode, and is specified to CGContextDrawPath directly
+            _figureInProgress = true;
+            CGPathMoveToPoint(_cgPath.get(), _transform, startPoint.x, -startPoint.y);
+        }
+    }
+
+    HRESULT STDMETHODCALLTYPE Close() {
+        if (_figureInProgress) {
+            TraceError(TAG,
+                       L"IDWriteGeometrySink::Close called while a figure was currently in progress. Placing object in error state - "
+                       L"future function calls will fail.");
+            _invalidState = true;
+        }
+
+        return _invalidState ? E_INVALIDARG : S_OK;
```

Comment:
It is essentially saying that the object wasn't expecting the call at this point.  E_FAIL is scarier.